pkg/geo/wkt/lex.go

```diff
@@ -275,43 +313,164 @@ func (l *wktLex) isValidStrideForLayout(stride int) bool {
 		return stride == 4
 	default:
 		// This should never happen.
-		panic("unknown geom.Layout")
+		panic(fmt.Sprintf("unknown geom.Layout %d", curLayout))
+	}
+}
+
+func (l *wktLex) isNonEmptyAllowedForLayout() bool {
+	if l.getCurLayoutMustBeEmpty() {
+		if l.getCurLayout() != geom.XYM {
+			panic("mustBeEmpty is true but layout is not XYM")
+		}
+		l.setIncorrectLayoutError(geom.NoLayout, "XYM layout must use the M variant of the geometry type")
+		return false
 	}
+	return true
 }
 
 func (l *wktLex) setLayout(layout geom.Layout) bool {
-	if layout == l.curLayout {
+	switch {
+	case layout == l.getCurLayout():
 		return true
-	}
-	if l.curLayout != geom.NoLayout {
+	case l.getCurLayout() != geom.NoLayout:
 		l.setIncorrectLayoutError(layout, "")
 		return false
+	default:
+		l.setLayoutIfNoLayout(layout)
+		return true
 	}
-	l.setLayoutIfNoLayout(layout)
+}
+
+func (l *wktLex) setLayoutBaseType() bool {
+	if !l.getCurLayoutBaseType() {
+		// Handle edge case where a GEOMETRYCOLLECTIONM may have base type EMPTY geometries but not non-EMPTY geometries.
+		if l.getCurLayout() == geom.XYM {
+			l.setCurLayoutMustBeEmpty(true)
+		}
+		return true
+	}
+
+	if l.getCurLayout() == geom.XYM {
+		l.setIncorrectLayoutError(geom.NoLayout, "XYM layout must use the M variant of the geometry type")
```

Comment:
nit: consider rewording this. it's not clear what is wrong (the XYM layout is a little confusing to an external user)